File: src/skillbank/agents.py
"""AgentConfig — agents.toml 加载 + per-agent 集成配置查询。"""

from __future__ import annotations

import dataclasses
import tomllib
from dataclasses import dataclass, field
from pathlib import Path
from typing import Optional

__all__ = ["AgentConfig", "AgentsConfig"]


@dataclass
class AgentConfig:
    """单个 Agent 的集成方式(对应 agents.toml 里 [agents.<name>])。

    字段语义见 agents.toml 注释:
- install_dir (仅文档参考! 部署路径唯一真相源是 machines.toml 的手填 skills_dir)
      / level 映射 / description_max / file_size_max / default_category / note

    全部 Agent 一律 cp 部署(7 个 emitter 均硬编码 cp; method 字段已删)。
    双语字段镜像(_zh <-> _cn)在对应 emitter 里硬编码, 不走配置。

    2026-08-15 决策:删除 resolve_install_dir(home 展开)。
    实测同一 Agent 跨机器路径不保证一致(QwenWorkCN), 显式手填 machines.toml 最稳;
    此处 install_dir 保留仅为 README 表格/文档参考, 代码不消费。
    """

    name: str
    display_name: str
    install_dir: str            # 仅文档参考, 部署不用它
    disable_invoke_field: Optional[str] = None
    disable_invoke_value: object = None
    description_max: Optional[int] = None
    file_size_max: Optional[int] = None
    default_category: Optional[str] = None
    skills_dir_config_key: Optional[str] = None
    note: str = ""

    def needs_disable_invoke(self, level_value: str) -> bool:
        """level == manual/experimental/disable 时, emitter 写出 disable_invoke_field"""
        return self.disable_invoke_field is not None and level_value in ("manual", "experimental", "disable")


@dataclass
class AgentsConfig:
    agents: dict[str, AgentConfig]

    @classmethod
    def load(cls, toml_path: Path) -> "AgentsConfig":
        with open(toml_path, "rb") as fh:
            d = tomllib.load(fh)
        agents = {}
        for name, body in d["agents"].items():
            agents[name] = AgentConfig(
                name=name,
                display_name=body.get("display_name", name),
                install_dir=body["install_dir"],
                disable_invoke_field=body.get("disable_invoke_field"),
                disable_invoke_value=body.get("disable_invoke_value"),
                description_max=body.get("description_max"),
                file_size_max=body.get("file_size_max"),
                default_category=body.get("default_category"),
                skills_dir_config_key=body.get("skills_dir_config_key"),
                note=body.get("note", ""),
            )
        return cls(agents=agents)

    def get(self, name: str) -> AgentConfig:
        return self.agents[name]

File: internal/config/agents.go
// Package config — agents.toml / machines.toml 加载 + skillbank 用户 config。
//
// agents.toml(对应 Python agents.py): 每 Agent 的集成方式。
// install_dir 仅文档参考, 部署路径唯一真相源是 machines.toml 手填 skills_dir。
package config

import (
	"os"

	"github.com/BurntSushi/toml"
)

// AgentConfig — 单个 Agent 的集成方式(agents.toml [agents.<name>])。
type AgentConfig struct {
	Name              string `toml:"-"`
	DisplayName       string `toml:"display_name"`
	InstallDir        string `toml:"install_dir"` // 仅文档参考, 部署不用
	DisableInvokeField string `toml:"disable_invoke_field"`
	DisableInvokeValue any    `toml:"disable_invoke_value"`
	DescriptionMax    int    `toml:"description_max"`  // 0 = 无限制
	FileSizeMax       int    `toml:"file_size_max"`    // 0 = 无限制
	DefaultCategory   string `toml:"default_category"`
	SkillsDirConfigKey string `toml:"skills_dir_config_key"`
	Note              string `toml:"note"`
}

// NeedsDisableInvoke — level 非 auto 时写 DisableInvokeField(空 = 该 Agent 无此字段, 如 kimi)。
func (c *AgentConfig) NeedsDisableInvoke(level string) bool {
	return c.DisableInvokeField != "" &&
		(level == "manual" || level == "experimental" || level == "disable")
}

// AgentsConfig — agents.toml 全表。
type AgentsConfig struct {
	Agents map[string]AgentConfig `toml:"agents"`
	// Order — agents.toml 声明序(Go map 无序, 用 toml MetaData 保留文档顺序;
	// sync 部署序 / list 列序跟随它, 与 Python dict 序等价)。
	Order []string `toml:"-"`
}

// LoadAgents — 读 agents.toml。
func LoadAgents(path string) (*AgentsConfig, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var cfg AgentsConfig
	md, err := toml.Decode(string(raw), &cfg)
	if err != nil {
		return nil, err
	}
	for name, a := range cfg.Agents {
		a.Name = name
		cfg.Agents[name] = a
	}
	// 保留 agents 声明序(文档顺序)
	seen := map[string]bool{}
	for _, key := range md.Keys() {
		if len(key) == 2 && key[0] == "agents" && !seen[key[1]] {
			seen[key[1]] = true
			cfg.Order = append(cfg.Order, key[1])
		}
	}
	return &cfg, nil
}

// Get — 按 agent 名取配置(不存在 panic 同 Python KeyError 语义, 调用方保证存在)。
func (ac *AgentsConfig) Get(name string) *AgentConfig {
	a, ok := ac.Agents[name]
	if !ok {
		return &AgentConfig{Name: name}
	}
	return &a
}

// Names — agents 声明序(agents.toml 文档顺序;缺 Order 时按字典序兜底)。
func (ac *AgentsConfig) Names() []string {
	if len(ac.Order) == len(ac.Agents) && len(ac.Order) > 0 {
		return ac.Order
	}
	out := make([]string, 0, len(ac.Agents))
	for name := range ac.Agents {
		out = append(out, name)
	}
	sortStrings(out)
	return out
}

func sortStrings(s []string) {
	for i := 1; i < len(s); i++ {
		for j := i; j > 0 && s[j] < s[j-1]; j-- {
			s[j], s[j-1] = s[j-1], s[j]
		}
	}
}